### PR TITLE
Step calculation based on CPU usage. 

### PR DESCRIPTION
**Pull request by RathnamVR**
## Purpose
Implementation of step limiting for choreo's milestone 2, measuring steps based on "How long does a service take to compute/complete?". In long running tasks and large payloads extra steps will be calculated for each additional 500ms of processing time. Epic wso2-enterprise/choreo#11966. 

## Goal
A new metric observer is published using choreo-extension for step counting purpose. This metric will calculate the total time consumed by a request using startobserver and stopobserver in observe.utils ballerina and do the step calculation based on that. At the end of a request there going to be a outgoing call to a public endpoint. The latency caused by this will not be deducted from total processing time because either way request is consuming CPU sources in that timespan too regardless the call being another choreo service or not. 

## Approach
Implemented a separate metric observer class name "StepcountObserver". At beginning of observation (observe.utils.startobserver) starting time will be recorded and at end of observation (observe.utils.stopobserver) end time will be taken to calculate the time spent on the process by taking the difference.
No of steps is then calculated using (duration-500)/500 and recorded on metric registry named "steps_total".
"StepcountObserver' is added to choreo-extension to be published on choreo.
Choreo telemetry- used to accept publishers from choreo-extension. Add "steps_total" metric registry to the list of metrics to be stored.

### CODE REVIEW

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            handleError("multiple metrics", tags, e);
+        }
+    }
+
+    private void handleError(String metricName, Set<Tag> tags, RuntimeException e) {
+        // Metric Provider may throw exceptions if there is a mismatch in tags.
+        consoleError.println("error: error collecting metrics for " + metricName + " with tags " + tags +
+                ": " + e.getMessage());
+    }
+}
```
Comment (tharinduwijewardane) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java:80]: Please add an empty line at the end of the file.

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+ */
+public class StepcountObserver implements BallerinaObserver {
+
+    private static final String PROPERTY_START_TIME = "_observation_start_time_";
+    
+    private static final PrintStream consoleError = System.err;
+
+    private static final MetricRegistry metricRegistry = DefaultMetricRegistry.getInstance();
+
+
```
Comment (tharinduwijewardane) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java:27]:
```suggestion

```

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/InitUtils.java [hunk 3]
```diff
@@ -25,10 +25,17 @@
...
 import io.ballerina.runtime.api.utils.StringUtils;
 import io.ballerina.runtime.api.values.BString;
+import io.ballerina.runtime.observability.ObserveUtils;
+import io.ballerina.runtime.observability.BallerinaObserver;
+
 
 import java.io.IOException;
 import java.util.Objects;
 
+
```
Comment (tharinduwijewardane) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/InitUtils.java:35]: Multiple consecutive empty lines are not recommended.

Comment (tharinduwijewardane) [reply choreo-extension-native/src/main/java/io/ballerina/observe/choreo/InitUtils.java:35]: Please fix the other places as well.

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
+package io.ballerina.observe.choreo;
```
Comment (tharinduwijewardane) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java:1]: You need to add the license header to each file (see other files)

**Comment by tharinduwijewardane** [comment]
Please read this article https://cbea.ms/git-commit

File: README.md [hunk 5]
```diff
@@ -1,6 +1,6 @@
 # Ballerina Choreo Observability Extension
 
-[![Build](https://github.com/ballerina-platform/module-ballerinax-choreo/workflows/Daily%20Build/badge.svg)](https://github.com/ballerina-platform/module-ballerinax-choreo/actions?query=workflow%3A"Daily+Build")
+[!  [Build](https://github.com/ballerina-platform/module-ballerinax-choreo/workflows/Daily%20Build/badge.svg)](https://github.com/ballerina-platform/module-ballerinax-choreo/actions?query=workflow%3A"Daily+Build")
```
Comment (nadundesilva) [inline README.md:3]: Let's revert this line. Seems to be an unintended change.

File: choreo-extension-ballerina/Ballerina.toml [hunk 6]
```diff
@@ -17,112 +17,112 @@
 [package]
 org = "ballerinax"
 name = "choreo"
-version = "@toml.version@"
+version = "0.4.10"
```
Comment (nadundesilva) [inline choreo-extension-ballerina/Ballerina.toml:20]: Ballerina.toml file should not be changed. Can you revert these changes as well ?

This normally reverts back to the original. But probably didn't due to it being stopped midway or some error 🤔

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/InitUtils.java [hunk 7]
```diff
@@ -40,14 +42,15 @@ private InitUtils() {   // Prevent initialization
     private static final String REPORTER_HOSTNAME_ENV_VAR = "CHOREO_EXT_REPORTER_HOSTNAME";
 
     private static final Logger LOGGER = LogFactory.getLogger();
-
+    private static final BallerinaObserver observer = new StepcountObserver();
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/InitUtils.java:45]: Better to add a new line between fields and methods.
```suggestion
    private static final BallerinaObserver observer = new StepcountObserver();

```

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+import io.ballerina.runtime.observability.metrics.MetricRegistry;
+import io.ballerina.runtime.observability.metrics.Tag;
+import io.ballerina.runtime.observability.metrics.DefaultMetricRegistry;
+import io.ballerina.runtime.observability.metrics.MetricId;
+
+/**
+ * Observe the runtime and collect measurements.
+ */
+public class StepcountObserver implements BallerinaObserver {
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java:35]:
```suggestion
public class StepCountObserver implements BallerinaObserver {
```

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    private static final String PROPERTY_START_TIME = "_observation_start_time_";
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java:36]: Let's not use the same key used in ballerina lang. They will conflict with each other and mess up the core metrics.

So let's prefix with `_choreo`
```suggestion
    private static final String PROPERTY_START_TIME = "_choreo_observation_start_time_";
```

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+            metricRegistry.counter(new MetricId("steps_total",
+                    "Total no of steps", tags)).increment(steps);
+        } catch (RuntimeException e) {
+            handleError("multiple metrics", tags, e);
+        }
+    }
+    
+    private void handleError(String metricName, Set<Tag> tags, RuntimeException e) {
+        // Metric Provider may throw exceptions if there is a mismatch in tags.
+        consoleError.println("error: error collecting metrics for " + metricName + " with tags " + tags +
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java:85]: For logging there is a separate logger in the extension. Check the `InitUtils.java` file. Let's use that.

Let's be consistent with the current code base always (unless there is something wrong with it and we need to fix it 😃).

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    }
+    
+    private void stopObservation(ObserverContext observerContext) {
+        Set<Tag> tags = new HashSet<>();
+        try {
+            Long startTime = (Long) observerContext.getProperty(PROPERTY_START_TIME);
+            long duration = System.nanoTime() - startTime;
+            long steps= Math.round(Math.ceil((double)(duration-500)/500));
+            //System.out.println(steps);
+            metricRegistry.counter(new MetricId("steps_total",
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java:76]: Let's add a prefix to prevent any possible future conflicts with other metrics
```suggestion
            metricRegistry.counter(new MetricId("choreo_steps_total",
```

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+                    "Total no of steps", tags)).increment(steps);
+        } catch (RuntimeException e) {
+            handleError("multiple metrics", tags, e);
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java:79]: We don't need a separate method for logging I think 🤔

In lang it is somewhat reused method. But here it does not make much sense to have a separate method I think.

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        observerContext.addProperty(PROPERTY_START_TIME, System.nanoTime());
+    }
+    
+    private void stopObservation(ObserverContext observerContext) {
+        Set<Tag> tags = new HashSet<>();
+        try {
+            Long startTime = (Long) observerContext.getProperty(PROPERTY_START_TIME);
+            long duration = System.nanoTime() - startTime;
+            long steps= Math.round(Math.ceil((double)(duration-500)/500));
+            //System.out.println(steps);
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java:75]: Should remove since this is an unused commented line.
```suggestion
```

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        }
+        stopObservation(observerContext);
+    }
+    
+    private void startObservation(ObserverContext observerContext) {
+        observerContext.addProperty(PROPERTY_START_TIME, System.nanoTime());
+    }
+    
+    private void stopObservation(ObserverContext observerContext) {
+        Set<Tag> tags = new HashSet<>();
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java:70]: Any reason to not add the tags from the observerContext ? 🤔 

```suggestion
        Set<Tag> tags = observerContext.getAllTags();
```

If we don't add there is no way to match back to the code I think 🤔

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        try {
+            Long startTime = (Long) observerContext.getProperty(PROPERTY_START_TIME);
+            long duration = System.nanoTime() - startTime;
+            long steps= Math.round(Math.ceil((double)(duration-500)/500));
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java:74]: Is this calculation correct ?

`System.nanoTime()` method returns the current time in nano seconds. Don't we need it in milliseconds ? 🤔
So shouldn't we be using `System.currentTimeMillis()` instead ?

Comment (sumedhassk) [reply choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java:74]: Can we take the value '500' from a configuration? (with 500 as default)

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java (hunk 10 above)
Comment (sumedhassk) [reply choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java:84]: Adding to what Nadun has mentioned, we normally use log4j for logging, and don't use console logs.

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+                    "Total no of steps", tags)).increment(steps);
+        } catch (RuntimeException e) {
+            handleError("multiple metrics", tags, e);
+        }
+    }
+    
+    private void handleError(String metricName, Set<Tag> tags, RuntimeException e) {
+        // Metric Provider may throw exceptions if there is a mismatch in tags.
+        consoleError.println("error: error collecting metrics for " + metricName + " with tags " + tags +
+                ": " + e.getMessage());
```
Comment (sumedhassk) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java:85]: Ideally we should log the full stacktrace if the exception is not propagated further

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java [hunk 17]
```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://wso2.com) All Rights Reserved.
```
Comment (sumedhassk) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java:2]: If this is a new file, this should be 2022

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java (hunk 15 above)
Comment (erandiganepola) [reply choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java]: +1 to read this from a config. As per Srinath, this value `500` should be verified after analysing a data set, they have used `500` as an example, not as an output of a analysis.

**Comment by erandiganepola** [comment]
Shall we squash and take multiple commits to a one/two commits?

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java (hunk 10 above)
Comment (nadundesilva) [reply choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepcountObserver.java]: There is a small blockers in using a logger like log4j.

We don't have a very good logger since this is within java native code within ballerina. The suggestion from ballerina team for this is to move everything to ballerina code which involves some amount of rewriting. But still they generally don't print java level stack traces AFAIK.

But in general we should follow what Sumedha suggested.

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.Set;
+
+/**
+ * Observe the runtime and collect measurements.
+ */
+
+public class StepCountObserver implements BallerinaObserver {
+    private static final String PROPERTY_START_TIME = "_choreo_observation_start_time_";
+
+    private static final String PROPERTY_STEP_TIME = System.getenv("choreo_per_step_time");
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java:37]: Shall we keep upper case env var name for consistency ?

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+import io.ballerina.runtime.observability.metrics.MetricId;
+import io.ballerina.runtime.observability.metrics.MetricRegistry;
+import io.ballerina.runtime.observability.metrics.Tag;
+
+import java.util.Set;
+
+/**
+ * Observe the runtime and collect measurements.
+ */
+
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java:33]: Seems like an unwanted additional empty line. Can u remove ?
```suggestion
```

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+        try {
+            Long startTime = (Long) observerContext.getProperty(PROPERTY_START_TIME);
+            long stepTime = Long.parseLong(PROPERTY_STEP_TIME);
+            long duration = System.currentTimeMillis() - startTime;
+            long steps = Math.round(Math.ceil((double) (duration - stepTime) / stepTime));
+            final MetricRegistry metricRegistry = DefaultMetricRegistry.getInstance();
+            metricRegistry.counter(new MetricId("choreo_steps_total", "Total no of steps", tags)).increment(steps);
+        } catch (RuntimeException e) {
+            LOGGER.error("error: error collecting metrics for multiple metrics with tags " 
+            + tags + ": " + e.getMessage());
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java:81]: Shall we exactly say what failed here in the error message ?
Also the `error: ` prefix is automatically added by the logger and therefore we should remove it from here.
```suggestion
            LOGGER.error("failed to record choreo_steps_total metric with tags " 
                + tags + ": " + e.getMessage());
```

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+    
+    private void startObservation(ObserverContext observerContext) {
+        observerContext.addProperty(PROPERTY_START_TIME, System.currentTimeMillis());
+    }
+    
+    private void stopObservation(ObserverContext observerContext) {
+        Set<Tag> tags = observerContext.getAllTags();
+        try {
+            Long startTime = (Long) observerContext.getProperty(PROPERTY_START_TIME);
+            long stepTime = Long.parseLong(PROPERTY_STEP_TIME);
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java:74]: We don't need to parse this everytime. Shall we move `Long.parseLong` call to the static field itself ?

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java (hunk 18 above)
Comment (nadundesilva) [reply choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java:37]:
```suggestion
    private static final String PROPERTY_STEP_TIME = System.getenv("choreo_per_step_time");
```

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+
+/**
+ * Observe the runtime and collect measurements.
+ */
+public class StepCountObserver implements BallerinaObserver {
+    private static final String PROPERTY_START_TIME = "_choreo_observation_start_time_";
+
+    private static final Logger LOGGER = LogFactory.getLogger();
+
+    private static long stepTime;
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java:38]: We can make this final.

```suggestion
    private static final long stepTime;
```

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+                propertysteptime = "500";
+            }
+    stepTime = Long.parseLong(propertysteptime);
+    }
+
+    @Override
+    public void startServerObservation(ObserverContext observerContext) {
+        startObservation(observerContext);
+    }
+    @Override
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java:52]: Please keep an empty line between two methods. Please do so in the methods below as well.
```suggestion
    }

    @Override
```

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+        Set<Tag> tags = observerContext.getAllTags();
+        try {
+            Long startTime = (Long) observerContext.getProperty(PROPERTY_START_TIME);
+            long duration = System.currentTimeMillis() - startTime;
+            long steps = Math.round(Math.ceil((double) (duration - stepTime) / stepTime));
+            final MetricRegistry metricRegistry = DefaultMetricRegistry.getInstance();
+            metricRegistry.counter(new MetricId("choreo_steps_total", "Total no of steps", tags)).increment(steps);
+        } catch (RuntimeException e) {
+            LOGGER.error("error collecting metrics for multiple metrics with tags " 
+            + tags + ": " + e.getMessage());
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java:87]: Generally we keep a tab when we go into multiple lines to indicate that it is from the previous line. It makes it easier to read when there is a lot of code.
```suggestion
                + tags + ": " + e.getMessage());
```

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+    private void stopObservation(ObserverContext observerContext) {
+        Set<Tag> tags = observerContext.getAllTags();
+        try {
+            Long startTime = (Long) observerContext.getProperty(PROPERTY_START_TIME);
+            long duration = System.currentTimeMillis() - startTime;
+            if (duration > stepTime) {
+                steps = Math.round(Math.ceil((double) (duration - stepTime) / stepTime));
+            }
+            final MetricRegistry metricRegistry = DefaultMetricRegistry.getInstance();
+            metricRegistry.counter(new MetricId("choreo_steps_total", "Total no of steps", tags)).increment(steps);
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java:92]: Keeping steps in a static variable will be problematic since there are multiple concurrent recordings happening. This will introduce concurrency issues as steps set by one thread will be overriden by another.

If we need to check whether duration > stepTime, we can just enclose everything within the if block.

```suggestion
            if (duration > stepTime) {
                long steps = Math.round(Math.ceil((double) (duration - stepTime) / stepTime));
                final MetricRegistry metricRegistry = DefaultMetricRegistry.getInstance();
                metricRegistry.counter(new MetricId("choreo_steps_total", "Total no of steps", tags)).increment(steps);
            }
```

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Override
+    public void stopServerObservation(ObserverContext observerContext) {
+        if (!observerContext.isStarted()) {
+            // Do not collect metrics if the observation hasn't started
+            return;
+        }
+        stopObservation(observerContext);
+
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java:65]:
```suggestion
```

File: choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+        Set<Tag> tags = observerContext.getAllTags();
+        try {
+            Long startTime = (Long) observerContext.getProperty(PROPERTY_START_TIME);
+            long duration = System.currentTimeMillis() - startTime;
+            if (duration > stepTime) {
+                long steps = Math.round(Math.ceil((double) (duration - stepTime) / stepTime));
+                final MetricRegistry metricRegistry = DefaultMetricRegistry.getInstance();
+                metricRegistry.counter(new MetricId("choreo_steps_total", "Total no of steps", tags)).increment(steps);
+            }
+
```
Comment (nadundesilva) [inline choreo-extension-native/src/main/java/io/ballerina/observe/choreo/StepCountObserver.java:91]:
```suggestion
```

**Review by nadundesilva — APPROVED** [review]

**Review by sumedhassk — APPROVED** [review]
